File: api/src/main/java/org/pmiops/workbench/google/CloudStorageService.java
package org.pmiops.workbench.google;

import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
import com.google.auth.oauth2.GoogleCredentials;
import com.google.cloud.storage.Blob;
import com.google.cloud.storage.BlobId;
import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.json.JSONObject;

/** Encapsulate Google APIs for interfacing with Google Cloud Storage. */
public interface CloudStorageService {

  String readInvitationKey();

  String readMandrillApiKey();

  String getImageUrl(String image_name);

  List<Blob> getBlobList(String bucketName);

  List<Blob> getBlobListForPrefix(String bucketName, String directory);

  Set<BlobId> getExistingBlobIdsIn(List<BlobId> id);

  void writeFile(String bucketName, String fileName, byte[] bytes);

  void copyBlob(BlobId from, BlobId to);

  JSONObject getJiraCredentials();

  JSONObject getElasticCredentials();

  GoogleCredentials getGSuiteAdminCredentials() throws IOException;

  GoogleCredentials getFireCloudAdminCredentials() throws IOException;

  GoogleCredentials getCloudResourceManagerAdminCredentials() throws IOException;

  GoogleCredentials getDefaultServiceAccountCredentials() throws IOException;

  GoogleCredentials getGarbageCollectionServiceAccountCredentials(String garbageCollectionEmail)
      throws IOException;

  Map<String, String> getMetadata(String bucketName, String objectPath);

  Blob getBlob(String bucketName, String objectPath);

  JSONObject readBlobAsJson(Blob blob);

  void deleteBlob(BlobId blobId);

  String getMoodleApiKey();
}

File: api/src/main/java/org/pmiops/workbench/google/CloudStorageService.kt
package org.pmiops.workbench.google

import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
import com.google.cloud.storage.Blob
import com.google.cloud.storage.BlobId
import java.io.IOException
import org.json.JSONObject

/** Encapsulate Google APIs for interfacing with Google Cloud Storage.  */
interface CloudStorageService {

    val jiraCredentials: JSONObject

    val elasticCredentials: JSONObject

    val gSuiteAdminCredentials: GoogleCredential

    val fireCloudAdminCredentials: GoogleCredential

    val cloudResourceManagerAdminCredentials: GoogleCredential

    val defaultServiceAccountCredentials: GoogleCredential

    val moodleApiKey: String

    fun readInvitationKey(): String

    fun readMandrillApiKey(): String

    fun getImageUrl(image_name: String): String

    fun getBlobList(bucketName: String): List<Blob>

    fun getBlobListForPrefix(bucketName: String, directory: String): List<Blob>

    fun blobsExist(id: List<BlobId>): Set<BlobId>

    fun writeFile(bucketName: String, fileName: String, bytes: ByteArray)

    fun copyBlob(from: BlobId, to: BlobId)

    @Throws(IOException::class)
    fun getGarbageCollectionServiceAccountCredentials(garbageCollectionEmail: String): GoogleCredential

    fun getFileAsJson(bucketName: String, fileName: String): JSONObject

    fun getMetadata(bucketName: String, objectPath: String): Map<String, String>

    fun deleteBlob(blobId: BlobId)
}
